File: elevator.py
class Elevator:
    def __init__(self, id, total_floors):
        self.id = id
        self.current_floor = 0
        self.target_floors = []
        self.direction = 'idle'
        self.queue = []
        self.total_floors = total_floors

    def move(self):
        if self.target_floors:
            if self.direction == 'up':
                self.current_floor += 1
            elif self.direction == 'down':
                self.current_floor -= 1

            if self.current_floor in self.target_floors:
                print(f"Elevator {self.id} stopping at floor {self.current_floor}")
                self.target_floors.remove(self.current_floor)

            if not self.target_floors and self.queue:
                self.target_floors.append(self.queue.pop(0))
                self.update_direction()

            if not self.target_floors:
               self.direction = 'idle'

    def add_request(self, floor):
        if floor not in self.target_floors and floor not in self.queue:
            if self.direction == 'idle' or ((self.direction == 'up' and floor > self.current_floor) or (self.direction == 'down' and floor < self.current_floor)):
                self.target_floors.append(floor)
                self.target_floors.sort(reverse=(self.direction == 'down'))
                self.update_direction()
            else:
                self.queue.append(floor)
            #If self.direction == 'idle':
            #    self.direction = 'up' if floor > self.current_floor else 'down'

    def update_direction(self):
        if self.target_floors:
            if self.target_floors[0] > self.current_floor:
                self.direction = 'up'
            elif self.target_floors[0] < self.current_floor:
                self.direction = 'down'

    def __str__(self):
        return (f"Elevator {self.id} | Current Floor: {self.current_floor} | "
                f"Direction: {self.direction} | Target Floors: {self.target_floors}")


class ElevatorSystem:
    def __init__(self, num_elevators, total_floors):
        self.elevators = [Elevator(i, total_floors) for i in range(num_elevators)]
        self.total_floors = total_floors

    def request_elevator(self, floor, direction):
        best_elevator = None
        min_distance = float('inf')
        for elevator in self.elevators:
            if elevator.direction == 'idle' or elevator.direction == direction:
                distance = abs(elevator.current_floor - floor)
                if distance < min_distance:
                    min_distance = distance
                    best_elevator = elevator
        if best_elevator:
            best_elevator.add_request(floor)
            return f'Elevator {best_elevator.id} assigned to floor {floor}'
        else:
            closest_elevator = min(self.elevators, key=lambda e: abs(e.current_floor - floor))
            closest_elevator.queue.append(floor)
            return f"Floor {floor} added to queue of Elevator {closest_elevator.id}"

    def step(self):
        for elevator in self.elevators:
            elevator.move()

    def display_status(self):
        for elevator in self.elevators:
            print(elevator)

# Example Usage
if __name__ == "__main__":
    # Initialize system with 3 elevators and 10 floors
    system = ElevatorSystem(num_elevators=1, total_floors=10)

    # Request elevators
    print(system.request_elevator(5, 'up'))
    print(system.request_elevator(3, 'down'))
    print(system.request_elevator(8, 'up'))
    print(system.request_elevator(1, 'down'))

    # Simulate the system
    print("\nInitial Status:")
    system.display_status()

    print("\nSimulating Movement:")
    for _ in range(10):  # Simulate 10 time steps
        system.step()
        system.display_status()
        print("-" * 40)

File: elevator.go
package main

import (
	"fmt"
	"sort"
)

type Elevator struct {
	ID           int
	CurrentFloor int
	TargetFloors []int
	Direction    string // "up", "down", or "idle"
	TotalFloors  int
}

func NewElevator(id, totalFloors int) *Elevator {
	return &Elevator{
		ID:          id,
		CurrentFloor: 0,
		Direction:   "idle",
		TotalFloors: totalFloors,
	}
}

func (e *Elevator) Move() {
	if len(e.TargetFloors) > 0 {
		if e.Direction == "idle" {
			e.updateDirection()
		}

		// Move one step in the current direction
		if e.Direction == "up" {
			e.CurrentFloor++
		} else if e.Direction == "down" {
			e.CurrentFloor--
		}

		// Check if the elevator has reached a target floor
		for i, floor := range e.TargetFloors {
			if floor == e.CurrentFloor {
				fmt.Printf("Elevator %d stopping at floor %d\n", e.ID, floor)
				e.TargetFloors = append(e.TargetFloors[:i], e.TargetFloors[i+1:]...)
				break
			}
		}

		// Update direction if no more target floors
		if len(e.TargetFloors) == 0 {
			e.Direction = "idle"
		}
	}
}

func (e *Elevator) AddTargetFloor(floor int) {
	// Avoid duplicate floor requests
	for _, target := range e.TargetFloors {
		if target == floor {
			return
		}
	}
	e.TargetFloors = append(e.TargetFloors, floor)
	e.sortTargetFloors()
	if e.Direction == "idle" {
		e.updateDirection()
	}
}

func (e *Elevator) updateDirection() {
	if len(e.TargetFloors) > 0 {
		if e.TargetFloors[0] > e.CurrentFloor {
			e.Direction = "up"
		} else if e.TargetFloors[0] < e.CurrentFloor {
			e.Direction = "down"
		}
	}
}

func (e *Elevator) sortTargetFloors() {
	if e.Direction == "down" {
		sort.Sort(sort.Reverse(sort.IntSlice(e.TargetFloors)))
	} else {
		sort.Ints(e.TargetFloors)
	}
}

func (e *Elevator) Status() string {
	return fmt.Sprintf("Elevator %d | Current Floor: %d | Direction: %s | Target Floors: %v",
		e.ID, e.CurrentFloor, e.Direction, e.TargetFloors)
}

type ElevatorSystem struct {
	Elevators   []*Elevator
	TotalFloors int
}

func NewElevatorSystem(numElevators, totalFloors int) *ElevatorSystem {
	elevators := make([]*Elevator, numElevators)
	for i := 0; i < numElevators; i++ {
		elevators[i] = NewElevator(i, totalFloors)
	}
	return &ElevatorSystem{
		Elevators:   elevators,
		TotalFloors: totalFloors,
	}
}

func (es *ElevatorSystem) RequestElevator(floor int) string {
	var bestElevator *Elevator
	minDistance := es.TotalFloors + 1

	for _, elevator := range es.Elevators {
		if elevator.Direction == "idle" || (elevator.Direction == "up" && floor >= elevator.CurrentFloor) ||
			(elevator.Direction == "down" && floor <= elevator.CurrentFloor) {
			distance := abs(elevator.CurrentFloor - floor)
			if distance < minDistance {
				minDistance = distance
				bestElevator = elevator
			}
		}
	}

	if bestElevator != nil {
		bestElevator.AddTargetFloor(floor)
		return fmt.Sprintf("Elevator %d assigned to floor %d", bestElevator.ID, floor)
	}

	// If no suitable elevator, add the request to the closest elevator
	closestElevator := es.Elevators[0]
	minDistance = abs(closestElevator.CurrentFloor - floor)
	for _, elevator := range es.Elevators {
		distance := abs(elevator.CurrentFloor - floor)
		if distance < minDistance {
			minDistance = distance
			closestElevator = elevator
		}
	}
	closestElevator.AddTargetFloor(floor)
	return fmt.Sprintf("Floor %d added to Elevator %d", floor, closestElevator.ID)
}

func (es *ElevatorSystem) Step() {
	for _, elevator := range es.Elevators {
		elevator.Move()
	}
}

func (es *ElevatorSystem) DisplayStatus() {
	for _, elevator := range es.Elevators {
		fmt.Println(elevator.Status())
	}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	// Initialize elevator system with 2 elevators and 10 floors
	system := NewElevatorSystem(2, 10)

	// Request elevators
	fmt.Println(system.RequestElevator(5)) // Elevator 0 should handle this
	fmt.Println(system.RequestElevator(3)) // Elevator 0 should handle this while moving to floor 5
	fmt.Println(system.RequestElevator(8)) // Elevator 1 should handle this
	fmt.Println(system.RequestElevator(1)) // Elevator 1 should handle this while moving to floor 8

	// Simulate elevator steps
	fmt.Println("\nInitial Status:")
	system.DisplayStatus()

	fmt.Println("\nSimulating Movement:")
	for i := 0; i < 10; i++ {
		system.Step()
		system.DisplayStatus()
		fmt.Println("----------------------------")
	}
}

